include/nbl/builtin/hlsl/emulated/float64_t_impl.hlsl

```diff
@@ -0,0 +1,312 @@
+#ifndef _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_IMPL_HLSL_INCLUDED_
+#define _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_IMPL_HLSL_INCLUDED_
+
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/ieee754.hlsl>
+#include <nbl/builtin/hlsl/algorithm.hlsl>
+#include <nbl/builtin/hlsl/tgmath.hlsl>
+#include <nbl/builtin/hlsl/glsl_compat/core.hlsl>
+
+// TODO: when it will be possible, use this unions wherever they fit:
+/*
+* union Mantissa
+* {
+*   struct
+*   {
+*       uint32_t highBits;
+*       uint64_t lowBits;
+*   };
+*
+*   uint32_t2 packed;
+* };
+*
+*/
+/*
+* union Mantissa
+* {
+*   struct
+*   {
+*       uint64_t lhs;
+*       uint64_t rhs;
+*   };
+*
+*   uint32_t4 packed;
+* };
+*
+*/
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace emulated_float64_t_impl
+{
+NBL_CONSTEXPR_INLINE_FUNC uint64_t2 shiftMantissaLeftBy53(uint64_t mantissa64)
+{
+    uint64_t2 output;
+    output.x = mantissa64 >> (64 - ieee754::traits<float64_t>::mantissaBitCnt);
+    output.y = mantissa64 << (ieee754::traits<float64_t>::mantissaBitCnt);
+
+    return output;
+}
+
+template<bool FlushDenormToZero>
+inline uint64_t castFloat32ToStorageType(float32_t val)
+{
+    if (FlushDenormToZero)
+    {
+        const uint64_t sign = uint64_t(ieee754::extractSign(val)) << 63;
+        if (tgmath::isInf(val))
+            return ieee754::traits<float64_t>::inf | sign;
+        uint32_t asUint = ieee754::impl::bitCastToUintType(val);
+        const int f32BiasedExp = int(ieee754::extractBiasedExponent(val));
+        if (f32BiasedExp == 0)
+            return sign;
+        const uint64_t biasedExp = uint64_t(f32BiasedExp - ieee754::traits<float32_t>::exponentBias + ieee754::traits<float64_t>::exponentBias) << (ieee754::traits<float64_t>::mantissaBitCnt);
+        const uint64_t mantissa = (uint64_t(ieee754::traits<float32_t>::mantissaMask) & asUint) << (ieee754::traits<float64_t>::mantissaBitCnt - ieee754::traits<float32_t>::mantissaBitCnt);
+
+        return sign | biasedExp | mantissa;
+    }
+    else
+    {
+        // static_assert(false);
+        return 0xdeadbeefbadcaffeull;
+    }
+};
+
+NBL_CONSTEXPR_INLINE_FUNC bool isZero(uint64_t val)
+{
+    return (val << 1) == 0;
+}
+
+// TODO: where do i move this function? also rename
+template <typename Int>
+inline int _findMSB(Int val)
+{
+    //static_assert(is_integral<Int>::value);
+#ifndef __HLSL_VERSION
+    return nbl::hlsl::findMSB(val);
+#else
+    return firstbithigh(val);
+#endif
+}
+
+template <>
+inline int _findMSB(uint64_t val)
+{
+#ifndef __HLSL_VERSION
+    return nbl::hlsl::findMSB(val);
+#else
+    int msbHigh = firstbithigh(uint32_t(val >> 32));
+    int msbLow = firstbithigh(uint32_t(val));
+    return msbHigh != -1 ? msbHigh + 32 : msbLow;
+#endif
+}
```

Comment:
should probably rewrite `hlsl::findMSB` as a template and specialize it as needed

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/696#discussion_r1748033737